routing/control_tower.go

```diff
@@ -205,82 +217,75 @@ func (p *controlTower) FetchInFlightPayments() ([]*channeldb.InFlightPayment, er
 // channel that provides the final outcome of the payment.
 func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
-	bool, chan PaymentResult, error) {
-
-	// Create a channel with buffer size 1. For every payment there will be
-	// exactly one event sent.
-	c := make(chan PaymentResult, 1)
+	*ControlTowerSubscriber, error) {
 
-	// Take lock before querying the db to prevent this scenario:
-	// FetchPayment returns us an in-flight state -> payment succeeds, but
-	// there is no subscriber to notify yet -> we add ourselves as a
-	// subscriber -> ... we will never receive a notification.
-	p.subscribersMtx.Lock()
-	defer p.subscribersMtx.Unlock()
+	// Take lock before querying the db to prevent missing or duplicating an
+	// update.
+	p.paymentsMtx.Lock(paymentHash)
+	defer p.paymentsMtx.Unlock(paymentHash)
 
 	payment, err := p.db.FetchPayment(paymentHash)
 	if err != nil {
-		return false, nil, err
+		return nil, err
 	}
 
-	var event PaymentResult
+	subscriber := newControlTowerSubscriber()
 
-	switch payment.Status {
+	// Always write current payment state to the channel.
+	subscriber.queue.ChanIn() <- payment
 
-	// Payment is currently in flight. Register this subscriber and
-	// return without writing a result to the channel yet.
-	case channeldb.StatusInFlight:
+	// Payment is currently in flight. Register this subscriber for further
+	// updates. Otherwise this update is the final update and the incoming
+	// channel can be closed. This will close the queue's outgoing channel
+	// when all updates have been written.
+	if payment.Status == channeldb.StatusInFlight {
+		p.subscribersMtx.Lock()
 		p.subscribers[paymentHash] = append(
-			p.subscribers[paymentHash], c,
-		)
-
-		return true, c, nil
-
-	// Payment already succeeded. It is not necessary to register as
-	// a subscriber, because we can send the result on the channel
-	// immediately.
-	case channeldb.StatusSucceeded:
-		event = *createSuccessResult(payment.HTLCs)
-
-	// Payment already failed. It is not necessary to register as a
-	// subscriber, because we can send the result on the channel
-	// immediately.
-	case channeldb.StatusFailed:
-		event = *createFailedResult(
-			payment.HTLCs, *payment.FailureReason,
+			p.subscribers[paymentHash], subscriber,
 		)
-
-	default:
-		return false, nil, errors.New("unknown payment status")
+		p.subscribersMtx.Unlock()
+	} else {
+		close(subscriber.queue.ChanIn())
 	}
 
-	// Write immediate result to the channel.
-	c <- event
-	close(c)
-
-	return false, c, nil
+	return subscriber, nil
 }
 
-// notifyFinalEvent sends a final payment event to all subscribers of this
+// notifySubscribers sends a final payment event to all subscribers of this
 // payment. The channel will be closed after this.
-func (p *controlTower) notifyFinalEvent(paymentHash lntypes.Hash,
-	event *PaymentResult) {
+func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
```

Comment:
add do doc that it should be called with the multimutex held?

Reply:
added